File: mobile/krishiai-app/src/database/models/Farm.ts
import { Model, Q } from '@nozbe/watermelondb';
import { field, date, relation, children } from '@nozbe/watermelondb/decorators';
import type { Associations } from '@nozbe/watermelondb/Model';
import type User from './User';
import type Crop from './Crop';

export default class Farm extends Model {
  static table = 'farms';
  static associations: Associations = {
    users: { type: 'belongs_to', key: 'user_id' },
    crops: { type: 'has_many', foreignKey: 'farm_id' },
  };

  @field('user_id') userId!: string;
  @field('location_lat') locationLat!: number;
  @field('location_lng') locationLng!: number;
  @field('size_hectares') sizeHectares!: number;
  @field('soil_type') soilType!: string;
  @field('irrigation_type') irrigationType!: string;
  @date('created_at') createdAt!: Date;
  @date('updated_at') updatedAt!: Date;

  @relation('users', 'user_id') user!: User;
  @children('crops') crops!: Q.Query<Crop>;
}

File: services/crop-service/src/models/farm.py
"""Farm database model."""

from sqlalchemy import Column, String, Numeric, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from ..database import Base


class Farm(Base):
    """Farm model representing farmer's land."""
    
    __tablename__ = "farms"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), nullable=False, index=True)
    
    # Location (stored as lat/lng, PostGIS GEOGRAPHY in production)
    latitude = Column(Numeric(10, 8), nullable=False)
    longitude = Column(Numeric(11, 8), nullable=False)
    
    # Farm characteristics
    size_hectares = Column(Numeric(10, 2), nullable=False)
    soil_type = Column(String(50), nullable=False)
    irrigation_type = Column(String(50), nullable=False)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    crops = relationship("Crop", back_populates="farm", cascade="all, delete-orphan")
    
    def __repr__(self):
        return f"<Farm(id={self.id}, user_id={self.user_id}, size={self.size_hectares}ha)>"
